### PR TITLE
feat: When a course is found in concluded/completed state, update instead of create ENT-4594

### PR DESCRIPTION
**Pull request by binodpant**
This change eliminates the failures encountered during create_content_metadata when a course is previously concluded (via a previous delete attempt). Instead, now we do a detection of course state before attempting to create it, and instead we update it if it's one of the following states in Canvas:
`'unpublished', 'completed' or 'available' `

- [x] This also brings a course back from completed to 'offer' state after which the course is now once again available for enrolling in Canvas
- [x] Also add detection of 'deleted' courses previously missed, since that allows us to log that a course has previojsly been 'deleted' as opposed to say 'concluded' (or 'completed'). We take no action on deleted courses other than logging for now
- [x] This can now even locate courses in other subaccount under the root account, as a fallback, in case not found under current subaccount.
- [x] Also, update_content_metadata and delete_content_metadata also use the same method I refactored `_get_course_id_from_edx_course_id` hence they will also benefit from being able to find a course in other subaccounts
- [x] Automated tests added. 

# Testing notes

After removing a course from catalog, verified it was in 'completed' workflow_state now. Then added back to catalog and ran transmit_content_metadata task and saw this in logs

```
2021-06-23 03:34:26,542 WARNING 49226 [integrated_channels.canvas.client] [user None] [ip None] client.py:129 - Course with canvas_id = 6072, integration_id = course-v1:edX+99887766+2T2021 found in workflow_state=completed,attempting to update instead of creating it
2021-06-23 03:34:27,026 WARNING 49226 [integrated_channels.canvas.client] [user None] [ip None] client.py:129 - Course with canvas_id = 6074, integration_id = edX+99887766 found in workflow_state=completed,attempting to update instead of creating it
2021-06-23 03:34:27,259 INFO 49226 [integrated_channels.integrated_channel.tasks] [user None] [ip None] tasks.py:47 - [Integrated Channel: CANVAS] Batch transmit_content_metadata finished in 1.3078796863555908 (api user: 3). Configuration: <CanvasEnterpriseCustomerConfiguration for Enterprise Test Enterprise>.
```

Then verified course in canvas went from ('concluded' or) 'completed' state to 'available'
Also verified via Canvas UI that course button now says 'Conclude this course', instead of 'Unconclude this course'

This payload obtained from `https://edx.instructure.com//api/v1/courses/6072?include=course_image&include=self_enrollment` shows the course state after the second round of transmit_content_metadata (there were no failures!)

```
{
    "id": 6072,
    "name": "bulk enroll test course",
    "account_id": 109,
    "uuid": "HkGD3EsTg8ihhoWArHlQRDMY8Aqy1hLqUfXUm62A",
    "start_at": "2021-05-26T16:00:00Z",
    "grading_standard_id": null,
    "is_public": true,
    "created_at": "2021-05-27T18:48:25Z",
    "course_code": "course-v1:edX+99887766+2T2021",
    "default_view": "syllabus",
    "root_account_id": 1,
    "enrollment_term_id": 1,
    "self_enrollment": true,
    "license": null,
    "grade_passback_setting": null,
    "end_at": "2022-08-27T16:00:00Z",
    "public_syllabus": false,
    "public_syllabus_to_auth": false,
    "storage_quota_mb": 500,
    "is_public_to_auth_users": false,
    "homeroom_course": false,
    "course_color": null,
    "hide_final_grades": false,
    "apply_assignment_group_weights": false,
    "calendar": {
        "ics": "https://edx.instructure.com/feeds/calendars/course_HkGD3EsTg8ihhoWArHlQRDMY8Aqy1hLqUfXUm62A.ics"
    },
    "time_zone": "America/Denver",
    "blueprint": false,
    "template": false,
    "sis_course_id": null,
    "sis_import_id": null,
    "integration_id": "course-v1:edX+99887766+2T2021",
    "enrollments": [],
    "workflow_state": "available",
    "restrict_enrollments_to_course_dates": false
}
```

__Also checked deleted course state handling__

Deleted a course in Canvas using the 'delete' (not conclude) option. Verified course returned now as 'workflow_state' = 'deleted' then ran a job (after clearing local tx records. This time instead of faiing to create it simply logged

```
2021-06-24 14:53:33,220 WARNING 51939 [integrated_channels.canvas.client] [user None] [ip None] client.py:122 - Course with integration_id = course-v1:edX+99887766+2T2021 found in deleted state,not attempting to create/update
```

__Finding course under root account instead of current subaccount__

I have to now add automated tests, but tested this out manually with various cases such as:

- deleted a course in canvas, then verified it can find it and log/ignore in that case (we still add a success tx record in this case)
- made a course unpublished in canvas, and it correctly detects state and updates course, instead of creating it
- forced it to try and find course under root account (by using a different canvas account id where course does not exist, and it did go find the course under root account id in that case)

```
2021-06-25 01:57:38,882 INFO 61047 [integrated_channels.canvas.client] [user None] [ip None] client.py:437 - [integrated_channel:canvas][integrated_channel_enterprise_customer_uuid:873fec72-5a21-4f95-ac19-e3d67f9c4777][integrated_channel_lms_user:None][integrated_channel_course_key:course-v1:edX+99887766+2T2021]Found course under root Canvas account
2021-06-25 01:57:38,883 ERROR 61047 [integrated_channels.canvas.client] [user None] [ip None] client.py:123 - [integrated_channel:canvas][integrated_channel_enterprise_customer_uuid:873fec72-5a21-4f95-ac19-e3d67f9c4777][integrated_channel_lms_user:None][integrated_channel_course_key:course-v1:edX+99887766+2T2021]Course with integration_id = course-v1:edX+99887766+2T2021 found in deleted state, not attempting to create/update
```

__Also I manually moved a course to a different subaccount in Canvas side__

Then remove the tx record to force it to create a course and voila! it found the course and instead updated it back from the concluded to offer state

ENT-4594

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.
- [ ] [Version](https://github.com/edx/edx-enterprise/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/edx/edx-enterprise/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/edx-enterprise/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/edx-enterprise/actions), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-platform](https://github.com/edx/edx-platform) to upgrade dependencies (including edx-enterprise)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: integrated_channels/canvas/client.py [hunk 1]
```diff
@@ -295,20 +353,38 @@ def _get_course_id_from_integration_id(self, integration_id):
...
         for course in all_courses_response:
-            if course['integration_id'] == integration_id:
-                course_id = course['id']
+            if course['integration_id'] == edx_course_id:
+                course_found = course
                 break
+        return course_found
 
-        if not course_id:
+        # now let's try under all subaccounts
```
Comment (binodpant) [inline integrated_channels/canvas/client.py:363]: Need to do this in case course was posted under a different subaccount / account previously

File: integrated_channels/canvas/client.py [hunk 2]
```diff
@@ -67,6 +70,16 @@ def course_create_endpoint(canvas_base_url, canvas_account_id):
             canvas_account_id,
         )
 
+    @staticmethod
+    def course_fetch_endpoint(canvas_base_url, canvas_account_id):
+        """
+        Returns endpoint to GET to for course fetch
```
Comment (edx-abolger) [inline integrated_channels/canvas/client.py:76]: nit: Can you make this docstring a bit more helpful?

File: integrated_channels/canvas/client.py [hunk 3]
```diff
@@ -67,6 +70,16 @@ def course_create_endpoint(canvas_base_url, canvas_account_id):
...
+        """
+        return '{}/api/v1/accounts/{}/courses'.format(
```
Comment (edx-abolger) [inline integrated_channels/canvas/client.py:78]: It looks like from here: https://canvas.instructure.com/doc/api/courses.html. that this API is an alias for the more straightforward GET /api/v1/courses/:id.   I would recommend you use this alias instead to keep the canvas account id out of the main logic and as part of the authentication setup only unless there is a good reason.

Comment (binodpant) [reply integrated_channels/canvas/client.py:78]: ah yes good callout. The only issue is that we don't have access to the canvas course id :( until we make a course fetch. This is something we should maybe think of doing, perhaps storing the canvas ids on our side would be nice as an association record. But we have not really put a place/table to do that as of yet. e.g. we could store the canvas course id + edx course id association in a table if we wanted sth like

```
customer_lms_configuration_id | edx_course_id | external_lms_course_id
```

or sth like it maybe as a future idea

File: integrated_channels/canvas/client.py (hunk 1 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:363]: I did not implement this part yet because I want to hear from reviewers what they think of this. Basically doing this means that we won't possibly be able to support multiple canvas accounts per customer., because we would be reaching out to find courses under all other subaccounts too. Meaning effecitvely merging in all subaccounts as if. they are all one set of courses (or ignoring the subaccounts in some way).

Should we only support one subaccount per customer? Should we try to find courses under other subaccounts?

With customers we don't yet have an issue since we only use one subaccount per customer so far. But our own demo instance uses multiple subaccounts. If we assume a course can only be posted under one subaccount, then we don't need to do this and we can simply acknowledge a failure to create a course in that case (after checking response of course)

Comment (binodpant) [reply integrated_channels/canvas/client.py:363]: Also, FYI it's easy to move course between subaccounts in Canvas (manually or via API), if needed

File: integrated_channels/canvas/client.py [hunk 4]
```diff
@@ -183,28 +234,33 @@ def delete_course_completion(self, user_id, payload):  # pylint: disable=unused-
...
-            if "image_url" in content_metadata_item:
-                url = CanvasAPIClient.course_update_endpoint(
-                    self.enterprise_configuration.canvas_base_url,
-                    course_id,
-                )
-                # Providing the param `event` and setting it to `offer` is equivalent to publishing the course.
-                self._put(url, json.dumps({
-                    'course': {'image_url': content_metadata_item['image_url'], 'event': 'offer'}
-                }).encode('utf-8'))
+            url = CanvasAPIClient.course_update_endpoint(
```
Comment (binodpant) [inline integrated_channels/canvas/client.py:252]: this refactor is here because we always want to set course to offer state, but the image_url detection of optional value is a separate concern, we don't want to mix the two

File: integrated_channels/canvas/client.py [hunk 5]
```diff
@@ -273,18 +331,26 @@ def _extract_integration_id(self, data):
...
+          skip these courses in logic for create_or_update when needed.
+
+          Note: we do not need to follow pagination here since it would be extremely unlikely
+          that searching by a specific edx_course_id results in many records, we generally only
+          expect 1 record to come back anyway.
+
+          The `&state[]=all` is added so we can also fetch priorly 'delete'd courses
         """
-        url = "{}/api/v1/accounts/{}/courses/?search_term={}".format(
+        url = "{}/api/v1/accounts/{}/courses/?search_term={}&state[]=all".format(
```
Comment (binodpant) [inline integrated_channels/canvas/client.py:350]: just later added this since without this param, 'deleted' courses will by default be excluded, and I have a clause elsewhere to specifically detect and log deleted courses

File: integrated_channels/canvas/client.py (hunk 3 above)
Comment (alex-sheehan-edx) [reply integrated_channels/canvas/client.py:78]: Do we need a course ID for that endpoint? I have a feeling that we can get all courses through `GET /api/v1/courses` but I'm not sure as to the scope of the courses returned. As in, if the `/api/v1/accounts/<account ID>/courses` endpoint actually limits to the sub-account or just returns all courses.

Comment (alex-sheehan-edx) [reply integrated_channels/canvas/client.py:78]: side note, are you using this method anywhere?

Comment (alex-sheehan-edx) [reply integrated_channels/canvas/client.py:78]: actually, assuming it's for https://github.com/edx/edx-enterprise/pull/1279/files#diff-bd39e800f794c671737477556e9d2d2e5855a2a4d5a83b7c764aba54ec49ece7R350 , I don't think we even need to worry about too many results since we're filtering by course title.

File: integrated_channels/canvas/client.py [hunk 6]
```diff
@@ -78,22 +91,62 @@ def course_update_endpoint(canvas_base_url, course_id):
         )
 
     def create_content_metadata(self, serialized_data):
+        """
+        Creates a course in Canvas.
+        If course is not found ,easy!  create it as usual
```
Comment (alex-sheehan-edx) [inline integrated_channels/canvas/client.py:96]: nit, I don't mind the `easy!` being there, but comma placement

File: integrated_channels/canvas/client.py [hunk 7]
```diff
@@ -78,22 +91,62 @@ def course_update_endpoint(canvas_base_url, course_id):
...
+        If course found, it will be one of these workflow_states per doc /the current state
```
Comment (alex-sheehan-edx) [inline integrated_channels/canvas/client.py:97]: `If course found, it will be one of these workflow_states per doc /the current state of the course one of 'unpublished', 'available', 'completed', or 'deleted'`

I'd maybe reword to something like
 
> If course is found, it will have one of the follow workflow_states: 
>                 `available`: issue an update with latest field values
>                 `completed`: this happens if a course has been concluded.
>                 ...
> for information of Canvas workflow_states see:
> https://canvas.instructure.com/doc/api/courses.html#method.courses.update

File: integrated_channels/canvas/client.py (hunk 1 above)
Comment (alex-sheehan-edx) [reply integrated_channels/canvas/client.py:371]: In with the convo going on with Adrienne's [comment](https://github.com/edx/edx-enterprise/pull/1279/files#r657913079), I'm wondering if we are able to hit the generic `GET /api/v1/courses` and get all courses in all sub-accounts. I think we should test out the functionality because it could simplify things.

File: integrated_channels/canvas/client.py [hunk 8]
```diff
@@ -183,28 +236,33 @@ def delete_course_completion(self, user_id, payload):  # pylint: disable=unused-
         pass
 
     # Private Methods
-
-    def _update_course_details(self, course_id, serialized_data):
+    def _update_course_details(self, course_id, course_details):
         """
         Update a course for image_url (and possibly other settings in future)
-        This is used only for settings that are not settable in the initial course creation
+        Also sets course to 'offer' state by sending 'course[event]=offer'
```
Comment (alex-sheehan-edx) [inline integrated_channels/canvas/client.py:242]: what does the offer state mean in this context? Does it have an effect on how we interact with the course?

File: integrated_channels/canvas/client.py [hunk 9]
```diff
@@ -183,28 +236,33 @@ def delete_course_completion(self, user_id, payload):  # pylint: disable=unused-
...
+
+        Arguments:
+          - course_details (dict): { 'image_url' } : optional used if present to course[image_url]
```
Comment (alex-sheehan-edx) [inline integrated_channels/canvas/client.py:245]: if you're going to add one of the params to the docstring I'd at least add the other even if it's straightforward as to not confuse anyone?

File: integrated_channels/canvas/client.py [hunk 10]
```diff
@@ -183,28 +236,33 @@ def delete_course_completion(self, user_id, payload):  # pylint: disable=unused-
...
         """
+        response_code = None
+        response_text = None
+        # Providing the param `event` and setting it to `offer` is equivalent to publishing the course.
```
Comment (alex-sheehan-edx) [inline integrated_channels/canvas/client.py:249]: I'd maybe include this in the docstring!

File: integrated_channels/canvas/client.py [hunk 11]
```diff
@@ -183,28 +236,33 @@ def delete_course_completion(self, user_id, payload):  # pylint: disable=unused-
...
-            if "image_url" in content_metadata_item:
-                url = CanvasAPIClient.course_update_endpoint(
-                    self.enterprise_configuration.canvas_base_url,
-                    course_id,
-                )
-                # Providing the param `event` and setting it to `offer` is equivalent to publishing the course.
-                self._put(url, json.dumps({
-                    'course': {'image_url': content_metadata_item['image_url'], 'event': 'offer'}
-                }).encode('utf-8'))
+            url = CanvasAPIClient.course_update_endpoint(
```
Comment (alex-sheehan-edx) [inline integrated_channels/canvas/client.py:254]: random nit, but the url generation is sitting in between two pieces of `update_payload` logic, maybe move this up above line 250 and keep things together?

**Review by alex-sheehan-edx — COMMENTED** [review]
Couple of comments and questions! But looking pretty good 😄

File: integrated_channels/canvas/client.py (hunk 6 above)
Comment (edx-abolger) [reply integrated_channels/canvas/client.py:96]: It boooooothers :P

File: integrated_channels/canvas/client.py [hunk 12]
```diff
@@ -78,22 +91,62 @@ def course_update_endpoint(canvas_base_url, course_id):
...
+            # Course does not exist: Create the course
+            status_code, response_text = self._post(self.course_create_url, serialized_data)
+            created_course_id = json.loads(response_text)['id']
+
+            # step 2: upload image_url and any other details
+            self._update_course_details(created_course_id, course_details)
+        else:
+            workflow_state = located_course['workflow_state']
+            if workflow_state.lower() == 'deleted':
+                LOGGER.warning(
```
Comment (edx-abolger) [inline integrated_channels/canvas/client.py:122]: Can you please use the new generate_formatted_log function for new logs in this file? user id will be none, but other items can be filled in.

Comment (edx-abolger) [reply integrated_channels/canvas/client.py:122]: Also can you raise this to "ERROR" ? If we have this it's going to cause sync problems for users, we want to know about it in the logs and possibly alert on it.

File: integrated_channels/canvas/client.py (hunk 8 above)
Comment (edx-abolger) [reply integrated_channels/canvas/client.py:242]: This could be resolved by adding a link to the relevant Canvas API.

File: integrated_channels/canvas/client.py [hunk 13]
```diff
@@ -183,28 +236,33 @@ def delete_course_completion(self, user_id, payload):  # pylint: disable=unused-
...
+            url = CanvasAPIClient.course_update_endpoint(
+                self.enterprise_configuration.canvas_base_url,
+                course_id,
+            )
+            if "image_url" in course_details:
+                update_payload['course']['image_url'] = course_details['image_url']
+
+            response_code, response_text = self._put(url, json.dumps(update_payload).encode('utf-8'))
         except Exception:  # pylint: disable=broad-except
             # we do not want course image update to cause failures
```
Comment (edx-abolger) [inline integrated_channels/canvas/client.py:263]: Please add an Error log here for alerting and tracking: I already have 2 different backlog tickets for bad course image URLs for other LMS systems, I don't want to add canvas to that list of problems I can't track.

File: integrated_channels/canvas/client.py [hunk 14]
```diff
@@ -295,20 +361,38 @@ def _get_course_id_from_integration_id(self, integration_id):
...
+        # now let's try under all subaccounts
+        """url = "{}/api/v1/accounts/{}/courses/?search_term={}".format(
+            self.enterprise_configuration.canvas_base_url,
+            self.enterprise_configuration.canvas_account_id,
+            quote(edx_course_id),
+        )"""
+
+    def _get_course_id_from_edx_course_id(self, edx_course_id):
+        """
+        To obtain course ID we have to request all courses associated with the integrated
```
Comment (edx-abolger) [inline integrated_channels/canvas/client.py:380]: This docstring scared me until I realized that it is not accurate. We do not get all courses and match ourselves, we call the remote API that does this. I would change this to say "To obtain the Canvas course ID, we have to use the Canvas API to search by edx_course_id."

**Review by edx-abolger — CHANGES_REQUESTED** [review]

File: integrated_channels/canvas/client.py (hunk 3 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:78]: I hear you @alex-sheehan-edx on the GET /api/v1/courses, but I checked it again and it only returns more specicifc results, which is weird. The doc also says only for this user. Meaning only courses where the current user (as in, which user? tracked by dev token I think), is enrolled or somehow associated So I could not get it to return the deleted or other course as returned by the search endpoint for example

I agree it would be nice to just call a generic, get me all courses, endpoint, I have not found one! still going to confirm

Comment (binodpant) [reply integrated_channels/canvas/client.py:78]: also checking this one https://canvas.instructure.com/doc/api/search.html#method.search.all_courses even though it says 'only courses in public index'

Comment (binodpant) [reply integrated_channels/canvas/client.py:78]: and yes, it does return courses across subaccounts! so this resolves one of the issues. The only issue is it does_not_return that deleted course :( if we are to accept that we won't see deleted courses (without using that other search method) and just not want to fetch 'deleted' courses, we can use this one .. I tried {{canvas_url}}/api/v1/search/all_courses?per_page=1000 and I get courses from multiple accounts...


Also if we use this endpoint which may return a large no. of courses, we may need to call it once per `client` lifecycle rather than once per call to `create_content_metadata`

File: integrated_channels/canvas/client.py (hunk 12 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:122]: both make sense! will do

File: integrated_channels/canvas/client.py (hunk 13 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:263]: sure, it will now look like this on error

```
021-06-24 18:18:20,270 ERROR 55966 [integrated_channels.canvas.client] [user None] [ip None] client.py:278 - [integrated_channel:canvas][integrated_channel_enterprise_customer_uuid:873fec72-5a21-4f95-ac19-e3d67f9c4777][integrated_channel_lms_user:None][integrated_channel_course_key:edX+99887766]Failed to update details for course, edx_course_id=edX+99887766, canvas_course_id=6074. Details: a test only
```

File: integrated_channels/canvas/client.py (hunk 3 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:78]: there may be one possible way to resolve this


we fetch the root account id (or even just have it stored in canvas config) : I can also try to fetch by calling /api/v1/accounts and using the one with parent_account_id=null.

if we have the root account id, then searching against that id, instead of the subaccount id like I do now, seems to find the course under various subaccount ids!

Comment (alex-sheehan-edx) [reply integrated_channels/canvas/client.py:78]: instead of fetching the root account ID, could we just ask the user to supply the root account ID within the configuration?

Comment (binodpant) [reply integrated_channels/canvas/client.py:78]: So after going over various options, here is a final idea I have to resolve all of this more cleanly:

```
def find_course_in_canvas(edx_course_id):
  course = find_course_in_account(current_subaccount)
  if not course:
     log(..)
      root_account = find_root_account()
      course = find_course_in_account(root_account) # this may or may not work even
 if not course:
    log(...)
  return course
```

with this, we will still mostly find courses under current subaccount (the most common case), but we will hit the root_account case but if we fail to find a course under current subaccount we will scour at a larger scope under the root account

Comment (edx-abolger) [reply integrated_channels/canvas/client.py:78]: @alex-sheehan-edx this config is already clearly too difficult for users as it is, I would much prefer we use the API call to get it if possible.

Comment (binodpant) [reply integrated_channels/canvas/client.py:79]: alrightie! added code to also check under root account id

File: integrated_channels/canvas/client.py [hunk 15]
```diff
@@ -358,33 +365,34 @@ def _extract_integration_id(self, data):
...
-            self.enterprise_configuration.canvas_account_id,
+            canvas_account_id,
             quote(edx_course_id),
         )
         resp = self.session.get(url)
         all_courses_response = resp.json()
 
         if resp.status_code >= 400:
+            message = ''
+            if 'reason' in all_courses_response:
```
Comment (binodpant) [inline integrated_channels/canvas/client.py:390]: it seems that when we hit a 404 we don't get a reason back, we just get 'errors' as an array

**Comment by binodpant** [comment]
there are a few test failures due to the changes to use a new utils file, will fix those up. A few mock changes are needed.

File: integrated_channels/canvas/client.py [hunk 16]
```diff
@@ -78,22 +94,85 @@ def course_update_endpoint(canvas_base_url, course_id):
...
         integration_id = self._extract_integration_id(serialized_data)
-        course_id = self._get_course_id_from_integration_id(integration_id)
+        course_id = CanvasUtil.get_course_id_from_edx_course_id(
+            self.enterprise_configuration,
+            self.session,
+            integration_id,
+        )
 
         url = CanvasAPIClient.course_update_endpoint(
             self.enterprise_configuration.canvas_base_url,
```
Comment (mattdrayer) [inline integrated_channels/canvas/client.py:178]: I recommend creating a getter for the `canvas_base_url` and ensuring that it appends a trailing slash to the end  of the value.  We definitely get burned on this from time to time, especially with SAPSF.

File: integrated_channels/canvas/utils.py [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+from integrated_channels.exceptions import ClientError
+from integrated_channels.utils import generate_formatted_log
+
+LOGGER = logging.getLogger(__name__)
+
+
+class CanvasUtil:
+    """
+    A util to make various util functions related to Canvas easier and co-located.
```
Comment (mattdrayer) [inline integrated_channels/canvas/utils.py:15]: Sorry, what's the point of having a static object, versus this simply being a library of functions?  It's not like we have to have objects for everything in Python-land.

Comment (binodpant) [reply integrated_channels/canvas/utils.py:15]: yeah good point, I had started off as a regular class with non static methods (since I had thought to do the fetch of root acount in the constructor and hold on to it), but for simplicity for now, I decided to go stateless, and I ended up with all static methods, and never thought to switch away from the class.

**Review by alex-sheehan-edx — APPROVED** [review]
I have the same questions/concern over base url as @mattdrayer  but otherwise this is looking good to me!

File: integrated_channels/canvas/client.py (hunk 16 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:178]: sure! I will take a look after pr to see if we manage that or not. but good callout, I think stripping makes sense to be safe to manual errors, e.g.

Comment (binodpant) [reply integrated_channels/canvas/client.py:178]: I can even do a micro pr for that

**Comment by binodpant** [comment]
@edx-abolger lmk if any concerns/flags, hoping ready to merge and test in staging today!

File: integrated_channels/canvas/utils.py (hunk 17 above)
Comment (binodpant) [reply integrated_channels/canvas/utils.py:15]: ever better, (in another PR), I think we want to use urllib.parse. 

```
>>> import urllib.parse
>>> urllib.parse.urljoin('http://test1','/abc')
'http://test1/abc'
>>> urllib.parse.urljoin('http://test1/','/abc')
'http://test1/abc'
>>> urllib.parse.urljoin('http://test1/','abc')
'http://test1/abc'
>>> urllib.parse.urljoin('http://test1','abc')
'http://test1/abc'
```

File: integrated_channels/canvas/client.py (hunk 16 above)
Comment (binodpant) [reply integrated_channels/canvas/client.py:178]: ever better, (in another PR), I think we want to use urllib.parse. 

```
>>> import urllib.parse
>>> urllib.parse.urljoin('http://test1','/abc')
'http://test1/abc'
>>> urllib.parse.urljoin('http://test1/','/abc')
'http://test1/abc'
>>> urllib.parse.urljoin('http://test1/','abc')
'http://test1/abc'
>>> urllib.parse.urljoin('http://test1','abc')
'http://test1/abc'
```

File: integrated_channels/canvas/client.py [hunk 18]
```diff
@@ -78,22 +94,85 @@ def course_update_endpoint(canvas_base_url, course_id):
...
 
-        # step 2: upload image_url and any other details
-        self._update_course_details(created_course_id, serialized_data)
+        if not located_course:
+            # Course does not exist: Create the course
+            status_code, response_text = self._post(self.course_create_url, serialized_data)
+            created_course_id = json.loads(response_text)['id']
+
+            # step 2: upload image_url and any other details
+            self._update_course_details(created_course_id, course_details)
```
Comment (edx-abolger) [inline integrated_channels/canvas/client.py:125]: Looks like we aren't even logging potential failures on this call. Can we at least add a log statement for failures here?

File: integrated_channels/utils.py [hunk 19]
```diff
@@ -247,7 +247,6 @@ def generate_formatted_log(
     Arguments:
     - channel_name (str): The name of the integrated channel
     - enterprise_customer_uuid (str): UUID of the relevant EnterpriseCustomer
-    - enterprise_customer_identifer (str): Identifying name of the EnterpriseCustomer
```
Comment (edx-abolger) [inline integrated_channels/utils.py:250]: Thanks for the catch here!

**Review by edx-abolger — APPROVED** [review]
LGTM at this time.

File: integrated_channels/utils.py (hunk 19 above)
Comment (binodpant) [reply integrated_channels/utils.py:250]: nw, I figured I would take care of this while I was here and the editor was pointing it out too, so great.
